src/consensus/validation.h

```diff
@@ -66,6 +66,32 @@ enum class ValidationInvalidReason {
     TX_MEMPOOL_POLICY,        //!< violated mempool's fee/size/descendant/RBF/etc limits
 };
 
+inline bool IsTransactionReason(ValidationInvalidReason r)
+{
+    return r == ValidationInvalidReason::NONE ||
+           r == ValidationInvalidReason::CONSENSUS ||
+           r == ValidationInvalidReason::RECENT_CONSENSUS_CHANGE ||
+           r == ValidationInvalidReason::TX_NOT_STANDARD ||
+           r == ValidationInvalidReason::TX_MISSING_INPUTS ||
+           r == ValidationInvalidReason::TX_WITNESS_MUTATED ||
+           r == ValidationInvalidReason::TX_CONFLICT ||
+           r == ValidationInvalidReason::TX_MEMPOOL_POLICY;
+}
+
+inline bool IsBlockReason(ValidationInvalidReason r)
```

Comment:
Oof this really sucks. To avoid delaying this PR I think this is fine, but we *really* need to just switch to two different ValidationState classes for Transactions and Blocks.

Reply:
FWIW, I had a go at this. See https://github.com/ajtowns/bitcoin/commits/201904-dos-rework-sep

I think it works okay, but splitting the type sure hits a lot of lines, see: https://github.com/ajtowns/bitcoin/commit/74bf5c12672d7eee877230d31b308c696fe83904 which is the commit that just splits the type. It's followed by the rest of this PR rebased on top of it, but with ValidationInvalidReason renamed to [Block|Tx]ValidationResult (since I'm touching every reference to that class anyway), and TX_PREMATURE_SPEND brought in initially rather than as a later commit.

Reply:
Nice work @ajtowns, but I think @sdaftuar might have a meltdown if we try to rebase this PR on any additional changes at this point. It seems to me that splitting `CValidationState` into separate classes for Tx and Block can be done in a follow-up PR to this. I'll commit to reviewing that follow-up PR if you open it after this is merged.

Reply:
Noting for later reference:

Definitely not worth changing in this PR now given # of acks, but this can probably be cleaned up in a later PR to just give each category of error a unique prefix (e.g, make all TX_* have MSByte set to 3 and all BLOCK_* have MSB set to 2 and general set to 1) or something.

```
enum class ValidationInvalidReason : uint64_t {
  IS_GENERAL REASON =1 << 56,
  // ....
  IS_BLOCK_REASON = 2 << 56
 // ....
  IS_TX_REASON = 3 << 56
}
```

Then these checks clean up to:

```
IS_TX_REASON & r || IS_GENERAL_REASON & r
```

The downside is depending on the enum grouping order, but then this makes it easier to later modify the code to split these enums into different classes (while preserving offsets), and then make the general wrapper functions take the correct type of enum.